sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,384 @@
+      return toBuilder().setFilenamePrefix(filenamePrefix).build();
+    }
+
+    /** The {@link CSVFormat} of the destination CSV file data. */
+    public Write<T> withCSVFormat(CSVFormat format) {
+      return toBuilder().setCSVFormat(format).build();
+    }
+
+    /**
+     * Not to be confused with the CSV header, it is content written to the top of every sharded
+     * file prior to the header. In the example below, all the text proceeding the header
+     * 'column1,column2,column3' is the preamble.
+     *
+     * <p>Fake company, Inc. Lab experiment: abcdefg123456 Experiment date: 2022-12-05 Operator:
+     * John Doe
+     *
+     * <p>column1,column2,column3 1,2,3 4,5,6
+     */
+    public Write<T> withPreamble(String preamble) {
+      return toBuilder().setPreamble(preamble).build();
+    }
+
+    /** Specifies the {@link Compression} of all generated shard files. */
+    public Write<T> withCompression(Compression compression) {
+      return toBuilder().setCompression(compression).build();
+    }
+
+    /**
+     * Specifies to use a given fixed number of shards per window. See {@link
+     * FileIO.Write#withNumShards(int)} for details.
+     */
+    public Write<T> withNumShards(Integer numShards) {
+      return toBuilder().setNumShards(numShards).build();
+    }
+
+    /**
+     * Specifies a directory into which all temporary files will be placed. See {@link
+     * FileIO.Write#withTempDirectory(String)}.
+     */
+    public Write<T> withTempDirectory(ResourceId value) {
+      return toBuilder().setTempDirectory(value).build();
+    }
+
+    abstract @Nullable String getFilenamePrefix();
+
+    abstract CSVFormat getCSVFormat();
+
+    abstract @Nullable String getPreamble();
+
+    abstract @Nullable Compression getCompression();
+
+    abstract @Nullable Integer getNumShards();
+
+    abstract String getFilenameSuffix();
+
+    abstract @Nullable ResourceId getTempDirectory();
+
+    abstract @Nullable List<String> getSchemaFields();
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+
+      abstract Builder<T> setFilenamePrefix(String value);
+
+      abstract Builder<T> setCSVFormat(CSVFormat value);
+
+      abstract Optional<CSVFormat> getCSVFormat();
+
+      abstract Builder<T> setPreamble(String value);
+
+      abstract Builder<T> setCompression(Compression value);
+
+      abstract Builder<T> setNumShards(Integer value);
+
+      abstract Builder<T> setFilenameSuffix(String value);
+
+      abstract Optional<String> getFilenameSuffix();
+
+      abstract Builder<T> setTempDirectory(ResourceId value);
+
+      abstract Builder<T> setSchemaFields(List<String> value);
+
+      abstract Write<T> autoBuild();
+
+      final Write<T> build() {
+        if (!getCSVFormat().isPresent()) {
+          setCSVFormat(CSVFormat.DEFAULT);
+        }
+
+        if (!getFilenameSuffix().isPresent()) {
+          setFilenameSuffix(DEFAULT_FILENAME_SUFFIX);
```

Comment:
Ditto for filename suffix to always start by setting our default and then allowing users to override it if they want.